src/crypto/slow-hash.c

```diff
@@ -59,35 +60,104 @@ extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *exp
+
+#define VARIANT2_INIT() \
+  uint32_t division_result[2]; \
+  uint32_t sqrt_result; \
+  do if (variant == 2) \
+  { \
+    *((uint64_t*)division_result) = 0; \
+    sqrt_result = 0; \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_SSE2(base_ptr, offset) \
+  do if (variant == 2) \
+  { \
+    const __m128i chunk1 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10))); \
+    const __m128i chunk2 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20))); \
+    const __m128i chunk3 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk3, _MM_SHUFFLE(2, 0, 3, 1)), _MM_SHUFFLE(3, 1, 2, 0))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk1, _MM_SHUFFLE(0, 3, 1, 2)), _MM_SHUFFLE(0, 3, 1, 2))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk2, _MM_SHUFFLE(3, 1, 2, 0)), _MM_SHUFFLE(2, 0, 3, 1))); \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_PORTABLE(base_ptr, offset) \
+  do if (variant == 2) \
+  { \
+    uint32_t* chunk1 = (uint32_t*)((base_ptr) + ((offset) ^ 0x10)); \
+    uint32_t* chunk2 = (uint32_t*)((base_ptr) + ((offset) ^ 0x20)); \
+    uint32_t* chunk3 = (uint32_t*)((base_ptr) + ((offset) ^ 0x30)); \
+    \
+    uint32_t chunk1_old[4]; \
+    memcpy(chunk1_old, chunk1, sizeof(chunk1_old)); \
+    \
+    chunk1[0] = (chunk3[1] & 0x0000FFFFUL) | (chunk3[3] << 16); \
+    chunk1[1] = (chunk3[3] & 0xFFFF0000UL) | (chunk3[1] >> 16); \
+    chunk1[2] = chunk3[0]; \
+    chunk1[3] = chunk3[2]; \
+    \
+    chunk3[0] = (chunk2[0] >> 16) | (chunk2[2] & 0xFFFF0000UL); \
+    chunk3[1] = (chunk2[2] << 16) | (chunk2[0] & 0x0000FFFFUL); \
+    chunk3[2] = chunk2[1]; \
+    chunk3[3] = chunk2[3]; \
+    \
+    chunk2[0] = (chunk1_old[2] & 0xFFFF0000UL) | (chunk1_old[1] & 0x0000FFFFUL); \
+    chunk2[1] = (chunk1_old[1] >> 16) | (chunk1_old[2] << 16); \
+    chunk2[2] = chunk1_old[3]; \
+    chunk2[3] = chunk1_old[0]; \
+  } while (0)
+
+#define VARIANT2_INTEGER_MATH_DIVISION_STEP(b, ptr) \
+  ((uint64_t*)(b))[1] ^= *((uint64_t*)division_result) ^ sqrt_result; \
+  division_result[0] = ((uint64_t*)(ptr))[1] / (((uint32_t*)(ptr))[0] | 0x80000001UL); \
+  division_result[1] = ((uint64_t*)(ptr))[1] % (((uint32_t*)(ptr))[0] | 0x80000001UL); \
+  const int64_t sqrt_input = (((uint64_t*)(ptr))[0] + *((uint64_t*)division_result)) >> 16;
+
+#define VARIANT2_INTEGER_MATH_SSE2(b, ptr) \
+  do if (variant == 2) \
+  { \
+    VARIANT2_INTEGER_MATH_DIVISION_STEP(b, ptr) \
+    const __m128d z = _mm_setzero_pd(); \
+    sqrt_result = _mm_cvttsd_si64(_mm_sqrt_sd(z, _mm_cvtsi64_sd(z, sqrt_input))); \
+  } while(0)
+
+#define VARIANT2_INTEGER_MATH_PORTABLE(b, ptr) \
+  do if (variant == 2) \
+  { \
+    VARIANT2_INTEGER_MATH_DIVISION_STEP(b, ptr) \
+    const double z = sqrt_input; \
+    const int64_t z1 = sqrt(z); \
```

Comment:
I did not realize this would actually use floating point. Is it stable across the whole 64 bit range on various platforms ?

Reply:
It's stable on any system that supports IEEE-754. Assuming that underlying hardware is IEEE-754 compliant, this code works like this:
- sqrt_input is 48-bit integer, it's converted to double without rounding and loosing precision
- sqrt(z) does rounding, but it's safe here: the result is less than 2^24 and absolute error is less than 2^-28 because **double** precision is used
- z is then truncated to get z1, but absolute error is too small to make this truncation incorrect

This was the theory, but I also tested this code (both portable and SSE2 versions) and it gives correct results for **all** integers less than 2^48.

I can publish the testing program so everyone else could confirm it. It's actually easy to write a testing program independently. The definition of integer square root is simple: INT_SQRT(N) is an integer number M such that both inequalities hold: M^2 <= N and (M+1)^2 > N. This number always exists if N >= 0 and there is always only one such number.

P.S. I can rewrite the fallback implementation (the last one in slow-hash.c) to use only integers to calculate the square root. It will become slower though.

Reply:
IEEE-754 behavior isn't defined by the C or C++ standard. There is a static check in `numeric_limits<T>` for IEEE. I doubt this check fails frequently though.

> * sqrt(z) does rounding, but it's safe here: the result is less than 2^24 and absolute error is less than 2^-28 because double precision is used
> * z is then truncated to get z1, but absolute error is too small to make this truncation incorrect

I might be missing something, but this feels awfully close to a +/-1 bug with rounding, where the float itself is `3.999999` or `4.00000`. I would need to brush up on my floating point behavior. @hyc ?

Reply:
IEEE-754 is not required, but even without it this code is far away to getting incorrect rounding: worst case is when sqrt(2^48 - 1) is calculated: https://onlinegdb.com/r1AmAyzrX

Rounding error is never larger than 1 ULP by definition, and the distance to the next integer is 16 ULP in this case.